.github/workflows/tests.yml

```diff
@@ -15,19 +15,28 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        python-version: ["2.7", "3.10", "pypy-2.7", "pypy-3.7"]
+        python-version:
+          - "2.7" # Python 2
+          - "3.7" # oldest supported Python 3
+          - "3.11" # newest supported Python 3
+          - "pypy-2.7" # PyPy 2
+          - "pypy-3.7" # oldest supported PyPy 3
+          - "pypy-3.9" # newest supported PyPy 3
 
     steps:
```

Comment:
Can you add something in here which actually uses mypy? So that we are actually getting some value out of the .pyi in the CI.

Reply:
That should be doable.

The `.py` files in `aocd` are ignored by `mypy` **because** the stubs exist, so perhaps we should check that the _tests_ use the type hints correctly?

I guess the other option would be to use the comment form of type hints directly within the `.py` files (except for `__init__.py`, where we can't (or at least, I don't know how to) due to the `sys.modules` hack), and then we could type check both the library and the tests (although, this'll be a larger change...).

There's also [`stubtest`](https://mypy.readthedocs.io/en/stable/stubtest.html), although I've never used it. All the other Python I've worked on using types is Python 3.5+, so I can just use the `variable_name: type` form in `.py` files (rather than `.pyi` stubs) so I've just type checked the library and tests directly with `mypy`.

It's also worth mentioning, `mypy` has fully dropped support for Python <= 3.5. Maybe we should only type check using Python 3.11 (at least, until you decide to drop support for older Python versions)?

Reply:
Yeah, it's fine to check only on one Python version & platform - whichever version is most convenient. Checking the tests seems OK, although I wouldn't be surprised if something in the mocking is not exactly type-correct (I am only considering "duck types" at best when writing test code)